### PR TITLE
Add the ruff linter to the pre-commit check

### PR DESCRIPTION
**Pull request by SamRWest**
I've been finding the [ruff linter](https://docs.astral.sh/ruff/rules/) really useful in some other projects, so thought I'd see if you guys were interested in adding it here.

My motivation was mainly to get the imports formatting standardised (which black doesn't do) for easier diffing/merging. 
But it's got a bunch of other really useful checks, especially when working with pandas (though I've turned these off for now so as not to confuse the PR diff).  

It's also found and fixed a bunch of style/safety things and I've flagged a couple of logic errors to be fixed also.

What do you guys think?

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -61,3 +62,37 @@ benchmark = { cmd = "python utils/run_benchmarks.py benchmarks.yml --run", help
...
+    "F", # pyflakes
+    "UP", # pyupgrade
+    "N", # pep8 naming
+    "I", # isort
+    "TID", # tidy imports
+    "UP", # pyupgrade
+    "NPY", # numpy style
+    "PL", # pylint
+#    "PD", # pandas style # TODO enable later
+#    "C90", # code complexity # TODO enable later
```
Comment (SamRWest) [inline pyproject.toml:86]: these last two will require a reasonable number of (simple) changes, so I've left them off for now so this diff isn't too polluted.

File: tests/test_transforms.py [hunk 2]
```diff
@@ -2,16 +2,16 @@
...
 import pandas as pd
 
-from xl2times import transforms, utils, datatypes
+from xl2times import datatypes, transforms, utils
 from xl2times.transforms import (
-    _process_comm_groups_vectorised,
     _count_comm_group_vectorised,
+    _match_wildcards,
+    _process_comm_groups_vectorised,
+    commodity_map,
```
Comment (SamRWest) [inline tests/test_transforms.py:10]: ^ This was the original motivation - standard import sorting/formatting

File: utils/dd_to_csv.py [hunk 3]
```diff
@@ -124,8 +123,8 @@ def parse_parameter_values_from_file(
 
 
 def save_data_with_headers(
-    param_data_dict: Dict[str, Union[pd.DataFrame, List[str]]],
-    headers_data: Dict[str, List[str]],
+    param_data_dict: dict[str, pd.DataFrame | list[str]],
```
Comment (SamRWest) [inline utils/dd_to_csv.py:126]: What python versions are you planning to support?
This was done with the `py11` setting (to match CI), but I'll revert and re-run for `py39` if that's the minimum target.  (I think the `|` type hint only works from 3.10 onwards)

File: utils/run_benchmarks.py [hunk 4]
```diff
@@ -65,6 +65,7 @@ def run_gams_gdxdiff(
         stdout=subprocess.PIPE,
         stderr=subprocess.STDOUT,
         text=True,
+        check=False,
```
Comment (SamRWest) [inline utils/run_benchmarks.py:68]: see https://docs.astral.sh/ruff/rules/subprocess-run-without-check/

File: utils/dd_to_csv.py (hunk 3 above)
Comment (olejandro) [reply utils/dd_to_csv.py:126]: We haven't had that discussion actually... How about `py12`?

**Comment by olejandro** [comment]
I generally like linters, but I don't have that much experience using them with python. I've used ESLint with javascript.

File: utils/dd_to_csv.py (hunk 3 above)
Comment (SamRWest) [reply utils/dd_to_csv.py:126]: Works for me, type hints are a lot cleaner in 3.12 for instance.
I just thought someone mentioned a use-case for running with an older version is all.

File: pyproject.toml [hunk 5]
```diff
@@ -61,3 +62,37 @@ benchmark = { cmd = "python utils/run_benchmarks.py benchmarks.yml --run", help
 benchmark_all = { shell = "python utils/run_benchmarks.py benchmarks.yml --verbose | tee out.txt", help = "Run the project", interpreter = "posix" }
 lint = { shell = "git add .pre-commit-config.yaml & pre-commit run", help = "Run pre-commit hooks", interpreter = "posix" }
 test = { cmd = "pytest --cov-report term --cov-report html --cov=xl2times --cov=utils", help = "Run unit tests with pytest" }
+
+
+# Config for various pre-commit checks are below
+# Ruff linting rules - see https://github.com/charliermarsh/ruff and https://beta.ruff.rs/docs/rules/
+[tool.ruff]
+target-version = "py39"
```
Comment (siddharth-krishna) [inline pyproject.toml:70]: I think we require Python >= 3.11 now, see line 13 of this file.

File: pyproject.toml (hunk 1 above)
Comment (siddharth-krishna) [reply pyproject.toml:86]: Nice idea, thanks. Looking forward to the next linting PR. :)

File: utils/dd_to_csv.py [hunk 6]
```diff
@@ -13,7 +12,7 @@
 
 def parse_parameter_values_from_file(
     path: Path,
-) -> Tuple[Dict[str, List], Dict[str, set]]:
+) -> tuple[dict[str, list], dict[str, set]]:
```
Comment (siddharth-krishna) [inline utils/dd_to_csv.py:15]: Am curious why the lowercase type annotations are preferable? (I'm happy either way)

File: utils/dd_to_csv.py (hunk 3 above)
Comment (siddharth-krishna) [reply utils/dd_to_csv.py:126]: I think we use the match-case statement somewhere in `transforms.py` which is >= 3.11. I feel perhaps we should try to keep the min version as low as possible, just to make it easier for users? Not sure how much of a pain it is to install a specific Python version in Windows (in linux I use `pyenv`)

File: utils/dd_to_csv.py [hunk 7]
```diff
@@ -201,12 +200,13 @@ def convert_dd_to_tabular(
...
     # Extract headers with key=param_name and value=List[attributes]
-    lines = list(open("xl2times/config/times_mapping.txt", "r"))
+    lines = list(open("xl2times/config/times_mapping.txt"))
     headers_data = headers_by_attr
     # The following will overwrite data obtained from headers_by_attr
     # TODO: Remove once migration is done?
     for line in lines:
-        line = line.strip()
+        # TODO Fix: PLW2901 `for` loop variable `line` overwritten by assignment target
+        line = line.strip()  # noqa
```
Comment (siddharth-krishna) [inline utils/dd_to_csv.py:209]: One way to fix this is to change the loop to `for line in (l.strip() for l in lines):`

File: xl2times/utils.py [hunk 8]
```diff
@@ -149,7 +150,8 @@ def apply_wildcards(
             current_list = []
             for wildcard in wildcard_list:
                 if wildcard.startswith("-"):
-                    wildcard = wildcard[1:]
+                    # TODO Fix: PLW2901 `for` loop variable `line` overwritten by assignment target
```
Comment (siddharth-krishna) [inline xl2times/utils.py:153]:
```suggestion
                    # TODO Fix: PLW2901 `for` loop variable `wildcard` overwritten by assignment target
```

**Review by siddharth-krishna — APPROVED** [review]
Great idea, thanks! I've been thinking of adding a linter for a while, though I haven't used ruff myself, it looks neat (and fast!).

I still have an open branch somewhere to upgrade the version of `pyright` -- this should also lead to a bunch of pandas-related changes, perhaps ruff will help us fix many of those?

**Review by olejandro — APPROVED** [review]

File: utils/dd_to_csv.py (hunk 3 above)
Comment (olejandro) [reply utils/dd_to_csv.py:126]: I'd say it is easy! :-) But I am also fine with sticking to 3.11.

File: utils/dd_to_csv.py (hunk 6 above)
Comment (SamRWest) [reply utils/dd_to_csv.py:15]: The lowercase hints allow the use of built-in types.  It avoid having to `from typing import Dict, List, ...` and lets you just hint with `dict, list` etc.  Just syntactic sugar.  Ruff updated all these automatically.

File: utils/dd_to_csv.py (hunk 3 above)
Comment (SamRWest) [reply utils/dd_to_csv.py:126]: Yeah I use pyenv on windows also, works very well.
Less savvy users can just download and run the official installers etc.
Ok, so we'll target >=3.11, cool.

**Review by siddharth-krishna — APPROVED** [review]

**Comment by olejandro** [comment]
@SamRWest I guess this one can be merged already?

**Comment by SamRWest** [comment]
> @SamRWest I guess this one can be merged already? Just so we don't have to resolve too many conflicts at a later stage...

Yep, I'll merge it shortly, then tackle #215

**Pull request merged**